README.md
# TutorialKit

[![Version](http://cocoapod-badges.herokuapp.com/v/TutorialKit/badge.png)](http://cocoadocs.org/docsets/TutorialKit)
[![Platform](http://cocoapod-badges.herokuapp.com/p/TutorialKit/badge.png)](http://cocoadocs.org/docsets/TutorialKit)

![alt tag](https://github.com/petrocket/TutorialKit/raw/master/Assets/tutorialkit.gif)

TutorialKit is a library for creating interactive step by step tutorials.  Highlight views with action messages, and show gestures without adding a lot of extra logic code to your app.


## Usage

**Simple example:**

- Add a tutorial sequence (the same one used in the gif above)
```Objective-C
    // Set up in the AppDelegate (but could be anywhere really)
    NSValue *msgPoint = [NSValue valueWithCGPoint:
                         CGPointMake(self.window.bounds.size.width * 0.5,
                                     self.window.bounds.size.height * 0.65)];
    NSValue *swipeStart = [NSValue valueWithCGPoint:
                           CGPointMake(self.window.bounds.size.width * 0.75,
                                       self.window.bounds.size.height * 0.75)];
    NSValue *swipeEnd = [NSValue valueWithCGPoint:
                           CGPointMake(self.window.bounds.size.width * 0.25,
                                       self.window.bounds.size.height * 0.75)];

    // set up a simple 3 step tutorial
    NSArray *steps = @[
                       // Step 0
                       @{
                           TKHighlightViewTag: @(1001), // tag assigned to your UIButton
                           TKMessage: @"First, press this button.",
                           TKMessagePoint: msgPoint
                           },
                       // Step 1
                       @{
                           TKSwipeGestureStartPoint: swipeStart,
                           TKSwipeGestureEndPoint: swipeEnd,
                           TKMessage: @"Next, swipe left.",
                           TKMessagePoint: msgPoint
                           },
                       // Step 2
                       @{
                           TKMessage: @"That's it! Yer all done!",
                           TKMessagePoint: msgPoint,
                           TKCompleteCallback: ^{ NSLog(@"ALL DONE."); }
                           },
                       ];
    
    [TutorialKit addTutorialSequence:steps name:@"example"];
```

- At any point in the application where you want to advance the tutorial:
```Objective-C
    [TutorialKit advanceTutorialSequenceWithName:@"example"];
```

- Profit.

**You can also configure the TutorialKit default styles.**
```Objective-C
    [TutorialKit setDefaultBlurAmount:0.5];
    [TutorialKit setDefaultMessageColor:UIColor.grayColor];
    [TutorialKit setDefaultMessageFont:[UIFont fontWithName:@"Helvetica" size:20]];
    [TutorialKit setDefaultTintColor:[UIColor colorWithWhite:1.0 alpha:0.5]];
``` 

**At any time you can set the tutorial step (useful if you want to start over)**
```Objective-C
    [TutorialKit setCurrentStep:0 forTutorial:@"example"];
```

To run the example project; clone the repo, and run `pod install` from the Example directory first.

## Installation

TutorialKit is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "TutorialKit"

## Author

Alex Peterson

## License

TutorialKit is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# TutorialKit CHANGELOG

## 0.1.9
Updated repository location

## 0.1.0

Initial release.


Classes/TutorialKit.m
/*
 TutorialKit.m
 Created by Alex on 4/21/14.
 Copyright (c) 2014 Alex Peterson. All rights reserved.
 
 The MIT License (MIT)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import "TutorialKit.h"
#import "UIView+Recursion.h"
#import "TutorialKitView.h"

#define TKSequence @"TutorialKitSequence"
#define TKStep @"TutorialKitStep"
#define TKUserDefaultsKey @"TutorialKitUserDefaults"

@interface TutorialKit()
@property (nonatomic, strong) NSMutableDictionary *sequences;
@property (nonatomic, strong) TutorialKitView *currentTutorialView;
@property (nonatomic, strong) UIColor *backgroundTintColor;
@property (nonatomic, strong) UIColor *labelColor;
@property (nonatomic, strong) UIFont *labelFont;
@property (nonatomic) CGFloat blurAmount;
@property (nonatomic) BOOL shouldContinue;
@end

@implementation TutorialKit

#pragma mark - Public

////////////////////////////////////////////////////////////////////////////////
+ (void)addTutorialSequence:(NSArray *)sequence name:(NSString*)name
{
    NSInteger step = [TutorialKit currentStepForTutorialWithName:name];;
    [TutorialKit.sharedInstance.sequences setObject:@{TKSequence:sequence,TKStep:@(step)} forKey:name];
}

////////////////////////////////////////////////////////////////////////////////
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name
{
    return [TutorialKit advanceTutorialSequenceWithName:name andContinue:NO];
}

////////////////////////////////////////////////////////////////////////////////
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name andContinue:(BOOL)shouldContinue
{
    if(TutorialKit.sharedInstance.currentTutorialView) {
        // a tutorial view is already visible
        return NO;
    }
    
    NSMutableDictionary *sequenceData = [TutorialKit.sharedInstance.sequences objectForKey:name];
    if(!sequenceData) {
        // assert?
        return NO;
    }
    
    if(![sequenceData isKindOfClass:NSMutableDictionary.class]) {
        sequenceData = sequenceData.mutableCopy;
    }
    
    NSNumber *step = [sequenceData objectForKey:TKStep];
    if(nil == step) {
        step = @(0);
        [sequenceData setObject:step forKey:TKStep];
    }
    
    NSArray *sequence = [sequenceData objectForKey:TKSequence];
    if(!sequence) {
        // assert?
        return NO;
    }
    
    if(step.integerValue >= sequence.count || step.integerValue < 0) {
        // sequence step is invalid or sequence is over
        return NO;
    }
    
    NSMutableDictionary *current = [sequence objectAtIndex:step.integerValue];
    if(!current) {
        // assert?
        return NO;
    }
    else if(![current isKindOfClass:NSMutableDictionary.class]) {
        // ensure this is a mutable copy
        current = current.mutableCopy;
    }

    if(![current objectForKey:TKMessageFont]) {
        [current setObject:TutorialKit.sharedInstance.labelFont forKey:TKMessageFont];
    }
    
    if(![current objectForKey:TKMessageColor]) {
        [current setObject:TutorialKit.sharedInstance.labelColor forKey:TKMessageColor];
    }
    
    if(![current objectForKey:TKBlurAmount]) {
        [current setObject:@(TutorialKit.sharedInstance.blurAmount) forKey:TKBlurAmount];
    }
    
    if([current objectForKey:TKHighlightViewTag]) {
        UIView *highlightView = [TutorialKit.sharedInstance
                                 findViewWithTag:[current objectForKey:TKHighlightViewTag]];
        if(highlightView) {
            [current setObject:highlightView forKey:TKHighlightView];
        }
        else {
            // highlight view not found! assert?
            return NO;
        }
    }

    TutorialKitView *tkv = [TutorialKitView tutorialViewWithDictionary:current];
    if(tkv) {
        tkv.sequenceStep = step.integerValue;
        tkv.sequenceName = name;
        tkv.tintColor = TutorialKit.sharedInstance.backgroundTintColor;
        UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc]
                                                 initWithTarget:TutorialKit.sharedInstance
                                                 action:@selector(dismissTutorialView:)];
        [tkv addGestureRecognizer:tapRecognizer];
        TutorialKit.sharedInstance.currentTutorialView = tkv;
        TutorialKit.sharedInstance.shouldContinue = shouldContinue;
        [TutorialKit presentTutorialView:tkv withAnimation:YES];
    }
    
    return tkv != NULL;
}

////////////////////////////////////////////////////////////////////////////////
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name ifOnStep:(NSInteger)step
{
    return [TutorialKit advanceTutorialSequenceWithName:name
                                               ifOnStep:step
                                            andContinue:NO];
}

////////////////////////////////////////////////////////////////////////////////
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name
                               ifOnStep:(NSInteger)step
                            andContinue:(BOOL)shouldContinue
{
    if([TutorialKit currentStepForTutorialWithName:name] != step) {
        return NO;
    }
    
    return [TutorialKit advanceTutorialSequenceWithName:name andContinue:shouldContinue];
}

////////////////////////////////////////////////////////////////////////////////
+ (NSInteger)currentStepForTutorialWithName:(NSString *)name
{
    NSMutableDictionary *sequence = [TutorialKit.sharedInstance.sequences objectForKey:name];
    if(sequence) {
        NSNumber *step = [sequence objectForKey:TKStep];
        if(step) return step.integerValue;
    }
    else {
        // check NSUserDefaults for a stored tutorial step
        NSInteger step = 0;
        if([NSUserDefaults.standardUserDefaults objectForKey:TKUserDefaultsKey]) {
            NSMutableDictionary *steps = [NSUserDefaults.standardUserDefaults objectForKey:TKUserDefaultsKey];
            if([steps objectForKey:name]) {
                step = [[steps objectForKey:name] integerValue];
            }
        }
        return step;
    }
    
    return 0;
}

////////////////////////////////////////////////////////////////////////////////
+ (UIView *)currentTutorialView
{
    return TutorialKit.sharedInstance.currentTutorialView;
}

////////////////////////////////////////////////////////////////////////////////
+ (void)dismissCurrentTutorialView
{
    if(TutorialKit.sharedInstance.currentTutorialView) {
        [TutorialKit.sharedInstance
         dismissTutorialView:TutorialKit.sharedInstance.currentTutorialView];
    }
}


////////////////////////////////////////////////////////////////////////////////
+ (void)insertTutorialSequence:(NSArray *)sequence name:(NSString*)name afterStep:(NSInteger)step
{
    [TutorialKit insertTutorialSequence:sequence name:name beforeStep:step + 1];
}

////////////////////////////////////////////////////////////////////////////////
+ (void)insertTutorialSequence:(NSArray *)sequence name:(NSString*)name beforeStep:(NSInteger)step
{
    NSDictionary *existingSequence = [TutorialKit.sharedInstance.sequences objectForKey:name];
    if(!existingSequence) {
        // @TODO Error message?
        return;
    }
    
    NSInteger curStep = [TutorialKit currentStepForTutorialWithName:name];
    NSArray *steps = existingSequence[TKSequence];
    if(step >= steps.count) {
        steps = [steps arrayByAddingObjectsFromArray:sequence];
    }
    else if(step > 0) {
        NSMutableArray *newSequence = @[].mutableCopy;
        [newSequence addObjectsFromArray:[steps subarrayWithRange:NSMakeRange(0, step)]];
        [newSequence addObjectsFromArray:sequence];
        [newSequence addObjectsFromArray:[steps subarrayWithRange:NSMakeRange(step, steps.count - step)]];
        steps = newSequence;
    }
    else {
        steps = [sequence arrayByAddingObjectsFromArray:steps];
    }
    
    [TutorialKit.sharedInstance.sequences setObject:@{TKSequence:steps,TKStep:@(curStep)} forKey:name];
}

////////////////////////////////////////////////////////////////////////////////
+ (void)presentTutorialView:(TutorialKitView *)view withAnimation:(BOOL)animate
{
    NSArray *windows = [UIApplication sharedApplication].windows;
    if(windows && windows.count > 0) {
        UIWindow *window = [windows objectAtIndex:0];
        [window addSubview:view];
        view.frame = window.bounds;
        [view setNeedsLayout];
        
        if(animate) {
            // fade in
            view.alpha = 0;
            [UIView animateWithDuration:0.5 animations:^{
                view.alpha = 1;
            } completion:nil];
        }
        else {
            view.alpha = 1.;
        }
    }
}

////////////////////////////////////////////////////////////////////////////////
+ (void)setCurrentStep:(NSInteger)step forTutorial:(NSString *)name
{
    NSMutableDictionary *sequence = [TutorialKit.sharedInstance.sequences objectForKey:name];
    if(sequence) {
        if(![sequence isKindOfClass:NSMutableDictionary.class]) {
            sequence = sequence.mutableCopy;
        }
        
        [sequence setObject:@(step) forKey:TKStep];
        
        // save to NSUserDefaults
        NSMutableDictionary *steps = [NSUserDefaults.standardUserDefaults objectForKey:TKUserDefaultsKey];
        if(!steps) {
            steps = @{}.mutableCopy;
        }
        else {
            steps = steps.mutableCopy;
        }
        [steps setObject:@(step) forKey:name];
        [NSUserDefaults.standardUserDefaults setObject:steps forKey:TKUserDefaultsKey];

        [TutorialKit.sharedInstance.sequences setObject:sequence forKey:name];
    }
}

////////////////////////////////////////////////////////////////////////////////
+ (void)setDefaultBlurAmount:(CGFloat)amount
{
    TutorialKit.sharedInstance.blurAmount = MAX(0.0,MIN(1.0,amount));
}

////////////////////////////////////////////////////////////////////////////////
+ (void)setDefaultMessageColor:(UIColor *)color
{
    TutorialKit.sharedInstance.labelColor = color;
}

////////////////////////////////////////////////////////////////////////////////
+ (void)setDefaultMessageFont:(UIFont *)font
{
    TutorialKit.sharedInstance.labelFont = font;
}

////////////////////////////////////////////////////////////////////////////////
+ (void)setDefaultTintColor:(UIColor *)color
{
    TutorialKit.sharedInstance.backgroundTintColor = color;
}

#pragma mark - Private

////////////////////////////////////////////////////////////////////////////////
- (UIColor *)backgroundTintColor
{
    if(!_backgroundTintColor) {
        _backgroundTintColor = [UIColor colorWithWhite:1.f alpha:0.5];
    }
    
    return _backgroundTintColor;
}

////////////////////////////////////////////////////////////////////////////////
- (void)dismissTutorialView:(TutorialKitView *)view
{
    if(view.sequenceName && view.sequenceStep >= 0) {
        // get the tutorial sequence with this name
        NSMutableDictionary *sequenceData = [TutorialKit.sharedInstance.sequences
                                             objectForKey:view.sequenceName];
        if(!sequenceData) {
            // assert?
            return;
        }
        else if(![sequenceData isKindOfClass:NSMutableDictionary.class]) {
            sequenceData = sequenceData.mutableCopy;
        }
        
        // increment the tutorial step
        [sequenceData setObject:@(view.sequenceStep + 1) forKey:TKStep];
        
        // save to NSUserDefaults
        NSMutableDictionary *steps = [NSUserDefaults.standardUserDefaults objectForKey:TKUserDefaultsKey];
        if(!steps) {
            steps = @{}.mutableCopy;
        }
        else {
            steps = steps.mutableCopy;
        }
        [steps setObject:@(view.sequenceStep + 1) forKey:view.sequenceName];
        [NSUserDefaults.standardUserDefaults setObject:steps forKey:TKUserDefaultsKey];
        
        // run the callback if one exists
        if(view.values && [view.values objectForKey:TKCompleteCallback]) {
            void (^callback)();
            callback = [view.values objectForKey:TKCompleteCallback];
            callback();
        }
        [TutorialKit.sharedInstance.sequences setObject:sequenceData
                                                 forKey:view.sequenceName];
    }
    if(view == TutorialKit.sharedInstance.currentTutorialView) {
        TutorialKit.sharedInstance.currentTutorialView = nil;
    }
    
    [UIView animateWithDuration:0.5 animations:^{
        view.alpha = 0;
    } completion:^(BOOL finished) {
        if(TutorialKit.sharedInstance.shouldContinue) {
            [TutorialKit advanceTutorialSequenceWithName:view.sequenceName];
        }
        else {
            [view removeFromSuperview];
        }
    }];
}

////////////////////////////////////////////////////////////////////////////////
- (UIView *)findViewWithTag:(NSNumber *)tag
{
    // look for the view with this tag
    __block UIView *tagView = nil;
    for(UIWindow *window in UIApplication.sharedApplication.windows) {
        [window.subviews enumerateObjectsUsingBlock:^(UIView *view, NSUInteger idx, BOOL *stop) {
            if(view.hidden || view.alpha == 0) return;
            
            tagView = [view findViewRecursively:^BOOL(UIView *subview, BOOL *stop) {
                if(subview.tag == tag.integerValue && !subview.hidden &&
                   subview.alpha != 0) {
                    *stop = YES;
                    return NO;
                }
                // return yes if should recurse further
                return !subview.hidden && subview.alpha != 0;
            }];
            
            if(tagView) *stop = YES;
        }];
        
        if(tagView) break;
    }
    
    return tagView;
}

////////////////////////////////////////////////////////////////////////////////
- (UIColor *)labelColor
{
    if(!_labelColor) {
        _labelColor = UIColor.grayColor;
    }
    return _labelColor;
}

////////////////////////////////////////////////////////////////////////////////
- (UIFont *)labelFont
{
    if(!_labelFont) {
        _labelFont = [UIFont fontWithName:@"Helvetica" size:16];
    }
    return _labelFont;
}

////////////////////////////////////////////////////////////////////////////////
- (void)onTapGesture:(UITapGestureRecognizer *)gesture
{
    if(!gesture.view || ![gesture.view isKindOfClass:TutorialKitView.class]) {
        // assert?
        return;
    }
    
    [self dismissTutorialView:(TutorialKitView *)gesture.view];
}

////////////////////////////////////////////////////////////////////////////////
- (NSMutableDictionary *)sequences
{
    if(!_sequences) {
        _sequences = @{}.mutableCopy;
    }
    return _sequences;
}

////////////////////////////////////////////////////////////////////////////////
+ (TutorialKit *)sharedInstance
{
    static TutorialKit *tk = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        if(!tk) tk = [[TutorialKit alloc] init];
        tk.currentTutorialView = nil;
        tk.blurAmount = 1.0;
    });
    
    return tk;
}

@end


Classes/TutorialKitView.h
/*
 TutorialKitView.h
 Created by Alex on 4/21/14.
 Copyright (c) 2014 Alex Peterson. All rights reserved.
 
 The MIT License (MIT)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#import <UIKit/UIKit.h>

@interface TutorialKitView : UIView

+ (instancetype) tutorialViewWithMessage:(NSString *)message
                           messageCenter:(CGPoint)messageCenter
                    messageCenterRelative:(BOOL)relativeMessageCenter
                                    font:(UIFont *)font
                                   color:(UIColor *)color
                           highlightView:(UIView *)view
                          highlightPoint:(CGPoint)point
                  highlightPointRelative:(BOOL)relativeHighlightPoint
                         highlightRadius:(float)radius;


+ (instancetype) tutorialViewWithMessage:(NSString *)message
                           messageCenter:(CGPoint)messageCenter
                   messageCenterRelative:(BOOL)relativeMessageCenter
                                    font:(UIFont *)font
                                   color:(UIColor *)color
                       swipeGestureStart:(CGPoint)start
                         swipeGestureEnd:(CGPoint)end
                   swipePositionsRelative:(BOOL)relativeSwipePositions
                         highlightRadius:(float)radius;

+ (instancetype) tutorialViewWithDictionary:(NSDictionary *)values;

@property (nonatomic, strong) NSDictionary *values;
@property (nonatomic, strong) NSString *sequenceName;
@property (nonatomic) NSInteger sequenceStep;
@property (nonatomic) CGFloat blurAmount;
@property (nonatomic, strong) UIColor *tintColor;

@end


Classes/UIView+Recursion.h
/*
 UIView+Recursion.h
 Created by Alex on 4/21/14.
 Copyright (c) 2014 Alex Peterson. All rights reserved.
 
 The MIT License (MIT)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#import <UIKit/UIKit.h>

@interface UIView (Recursion)

/** Find a UIView recursively.  Return TRUE from the block to recurse into subview.
 Set stop to TRUE to stop recursing and return the subview.

 @param recurse The block to use to determine whether to continue recursing
 @return Returns the UIView if found or nil
 */
- (UIView*)findViewRecursively:(BOOL(^)(UIView* subview, BOOL* stop))recurse;

@end


Classes/UIView+Recursion.m
/*
 UIView+Recursion.m
 Created by Alex on 4/21/14.
 Copyright (c) 2014 Alex Peterson. All rights reserved.
 
 The MIT License (MIT)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import "UIView+Recursion.h"

@implementation UIView (Recursion)

////////////////////////////////////////////////////////////////////////////////
- (UIView*)findViewRecursively:(BOOL(^)(UIView* subview, BOOL* stop))recurse
{
    for( UIView* subview in self.subviews ) {
        BOOL stop = NO;
        if( recurse( subview, &stop ) ) {
            UIView *view = [subview findViewRecursively:recurse];
            if(view) return view;
        } else if( stop ) {
            return subview;
        }
    }
    return nil;
}

@end


Classes/TutorialKit.h
/*
 TutorialKit.h
 Created by Alex on 4/21/14.
 Copyright (c) 2014 Alex Peterson. All rights reserved.
 
 The MIT License (MIT)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import <Foundation/Foundation.h>

static NSString* const TKBlurAmount = @"TKBlurAmount";
static NSString* const TKMessage = @"TKMessage";
static NSString* const TKMessageColor = @"TKMessageColor";
static NSString* const TKMessageFont = @"TKMessageFont";
static NSString* const TKMessagePoint = @"TKMessagePoint"; // absolute 0..width, 0..height
static NSString* const TKMessageRelativePoint = @"TKMessageRelativePoint"; // relative 0..1, 0..1
static NSString* const TKHighlightView = @"TKHighlightView";
static NSString* const TKHighlightViewTag = @"TKHighlightViewTag";
static NSString* const TKHighlightPoint = @"TKHighlightPoint"; // absolute 0..width, 0..height
static NSString* const TKHighlightRelativePoint = @"TKHighlightRelativePoint";  // relative 0..1
static NSString* const TKHighlightRadius = @"TKHighlightRadius";
static NSString* const TKSwipeGestureStartPoint = @"TKSwipeGestureStartPoint";
static NSString* const TKSwipeGestureRelativeStartPoint = @"TKSwipeGestureRelativeStartPoint";
static NSString* const TKSwipeGestureEndPoint = @"TKSwipeGestureEndPoint";
static NSString* const TKSwipeGestureRelativeEndPoint = @"TKSwipeGestureRelativeEndPoint";
static NSString* const TKCompleteCallback = @"TKCompleteCallback";

@interface TutorialKit : NSObject

/** Adds a tutorial sequence to display
 
 Example:
 [self addTutorialSequence:@[
 // highlight a view and display a message
 @{
 TKMessage:@"first message",
 TKMessagePoint:[NSValue valueWithCGPoint:CGPointMake(100,100)],
 TKHighlightViewTag:1,
 TKCompleteCallback:^{ NSLog("Step 1 complete!"); }
 },
 
 // highlight a specific point in a view and display a message
 @{
 TKMessage:@"second message",
 TKMessageRelativePoint:[NSValue valueWithCGPoint:CGPointMake(0.5,0.75)],
 TKHighlightViewTag:1,
 TKHighlightRelativePoint:[NSValue valueWithCGPoint:CGPointMake(0.5,0.5)],
 TKHighlightRadius:@(100)
 TKCompleteCallback:^{ NSLog("Sequence complete!"); }
 },
 
 ]];
 
 
 @param array of tutorial sequence dictionaries
 @param name The name of this tutorial sequence
 */
+ (void)addTutorialSequence:(NSArray *)sequence name:(NSString*)name;

/** Advance the tutorial sequence is possible and does not auto-continue
 
 @param name The name of the tutorial sequence to advance
 @return Returns TRUE if the tutorial sequence advanced
 */
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name;

/** Advance the tutorial sequence is possible
 
 @param name The name of the tutorial sequence to advance
 @param continue True if should continue to next tutorial step if possible
 @return Returns TRUE if the tutorial sequence advanced
 */
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name
                            andContinue:(BOOL)shouldContinue;

/** Advance the tutorial sequence is possible and only if on this step and does
 not auto-continue.
 
 @param name The name of the tutorial sequence to advance
 @param step Only advance if the tutorial is on this step
 @return Returns TRUE if the tutorial sequence advanced
 */
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name
                               ifOnStep:(NSInteger)step;

/** Advance the tutorial sequence is possible and only if on this step and
 continue if requested.
 
 @param name The name of the tutorial sequence to advance
 @param step Only advance if the tutorial is on this step
 @param shouldContinue True if should continue to next tutorial step if possible
 @return Returns TRUE if the tutorial sequence advanced
 */
+ (BOOL)advanceTutorialSequenceWithName:(NSString *)name
                               ifOnStep:(NSInteger)step
                            andContinue:(BOOL)shouldContinue;

/** Current step for the specified tutorial
 
 @param name The name of the tutorial sequence
 @return Returns the current step for the specified tutorial
 */
+ (NSInteger)currentStepForTutorialWithName:(NSString *)name;


/** The UIView for the active tutorial step
 
 @return Returns the current tutorial UIView or nil if no tutorial is visible
 */
+ (UIView *)currentTutorialView;

/** Dismiss the current tutorial view
 */
+ (void)dismissCurrentTutorialView;

/** Inserts a tutorial sequence into the existince sequence
 @param array of tutorial sequence dictionaries
 @param name The name of this tutorial sequence
 @param name The name of this tutorial sequence
 @param step The step to after which to insert this sequence
 */
+ (void)insertTutorialSequence:(NSArray *)sequence name:(NSString*)name afterStep:(NSInteger)step;

/** Inserts a tutorial sequence into the existince sequence
 @param array of tutorial sequence dictionaries
 @param name The name of this tutorial sequence
 @param name The name of this tutorial sequence
 @param step The step to before which to insert this sequence
 */
+ (void)insertTutorialSequence:(NSArray *)sequence name:(NSString*)name beforeStep:(NSInteger)step;

/** Set the current step for the specified tutorial
 
 @param step The current step for the specified tutorial sequence
 @param name The name of the tutorial sequence
 */
+ (void)setCurrentStep:(NSInteger)step forTutorial:(NSString *)name;

/** Set the default blur amount.  Set to zero to disable blur
 
 @param color The amount to blur the underlying image
 */
+ (void)setDefaultBlurAmount:(CGFloat)amount;

/** Set the default message color
 
 @param color The UIColor to use for messages by default
 */
+ (void)setDefaultMessageColor:(UIColor *)color;

/** Set the default font to use
 
 @param font The UIFont to use for messages by default
 */
+ (void)setDefaultMessageFont:(UIFont *)font;

/** Set the default background tint color
 
 @param color The UIColor to use for the background tint
 */
+ (void)setDefaultTintColor:(UIColor *)color;

@end
